### DISCUSSION TITLE
Is replication sending supposed to be heavy when it can't send?

### DISCUSSION THREAD

**Discussion by Gomez12** [ravendb/ravendb#18215] [Q&A]
We have a cluster of 3 nodes running 5.4.116

Because of an error I had to hard reset our main database on one node (this takes approx 12+ hours)
This is not an issue, but this node is now currently running like 50+% cpu load while reindexing the main database.
Which seems to mean that secondary databases are not sending replication documents to this node (not a problem this far, as the secondary databases still have 2 nodes to respond)

But if a secondary database now receives a new document it wants to replicate this to the other nodes, one node goes ok. But it can't send it to the node which is doing heavy reindexing, but just trying to replicate it to this node puts a 50% cpu load on the sending node.

Or with node names A/B/C
node C is doing heavy reindexing work on the main db which puts the cpu load to 50+% on this node.
node B receives a new document on a secondary database, it immediately replicates this to node A (no problem), but in threads runtime info it says that it is using 40% cpu with : outgoing replication b->c on secondary database.

If I now pauze the reindexing on node C, then the replication (on the secondary database) from node B goes through again, and the load on node B goes down to like 5%, If I resume the reindexing on node C again, then the replication stops again (which is no problem) but the cpu load on node B goes to 40% again on the outgoing replication.

Is there any way to get the replication to back off if it can't send it, because currently 2 nodes are having heavy-cpu uses because I am reindexing 1 node.

The secondary database is a logging-database, and node C is currently about 10 million documents behind node B/A

**Comment by Gomez12** [comment]
Ok, the problem is less than it seemed. The cpu load on B is mostly coming from the no. of documents to replicate.

I have now paused the reindexing on node C to allow node B to fully replicate with node C (previous I would let about 1 million documents replicate and then start indexing again). So all the nodes where equal on the secondary database.
And now node C is indexing again.

I have a backlog for replication which is going up on B, but it doesn't take up 40% cpu as long as it is a small backlog it seems.
So maybe you can optimize it a bit more, but just not letting the backlog run too high also solves the issue.

**Comment by ayende** [comment]
The actual reason isn't about CPU, but I/O load, most likely.
Basically, during indexing, especially at 50% CPU, we are writing a lot to the disk.
That means that replication on the secondary db will compete for the same disk resources. It can cause the replication task to fail with a timeout because the replication batch will take a while to commit to the disk.

That is handled, by the way, and retries will grow in exponential manner until the load is reduced.
